### PR TITLE
perf(scan): optimize writing of scanned images to disk

### PR DESCRIPTION
**Pull request by eventualbuddha**
## Overview

Previously we were writing both front & back images to disk twice: once before interpretation (as a kind of staging area, I guess?) and once after. This removes the first write (and the subsequent read+decode). In addition, this ensures that the PNG encoding (which is the slowest part of saving the images) happens in parallel.

## Demo Video or Screenshot
Testing on my Core i5, I'm seeing the time spend just encoding images and writing to disk going from ~450ms down to 250-300ms. Here's a chart of the total time for _all of intake, transfer, interpret, write and record_ before and after:

image

The median scan time went from 3.47s to 2.865s, a decrease of about 23%.

## Testing Plan
Manual testing, new automated testing to prevent regression of the PNG encoding parallelism.

### CODE REVIEW

File: apps/scan/backend/src/scanners/pdi/state_machine.ts [hunk 1]
```diff
@@ -19,52 +29,33 @@ import {
...
   PrecinctScannerStateMachine,
 } from '../../types';
+import { rootDebug } from '../../util/debug';
+import { Workspace } from '../../util/workspace';
 import { recordAcceptedSheet, recordRejectedSheet } from '../shared';
-import { isReadyToScan } from '../../app_flow';
 
 const debug = rootDebug.extend('state-machine');
 
+let scanAndInterpretTimer: Timer | undefined;
```
Comment (jonahkagan) [inline apps/scan/backend/src/scanners/pdi/state_machine.ts:47]: Just making sure you want to leave the timer in

File: libs/image-utils/src/image_data.ts [hunk 2]
```diff
@@ -96,38 +101,103 @@ export function fromGrayScale(
...
+  const canvas = createCanvasWithImageData(image);
   return mimeType === 'image/jpeg'
     ? canvas.toDataURL(mimeType)
     : canvas.toDataURL(mimeType);
 }
 
 /**
  * Writes an image to a file.
+ *
+ * @param buffer - Whether to buffer the image before writing it to disk.
```
Comment (jonahkagan) [inline libs/image-utils/src/image_data.ts:127]: I don't see a `buffer` param - is this outdated?

**Review by jonahkagan — APPROVED** [review]
Long time coming!

File: libs/image-utils/src/image_data.ts (hunk 2 above)
Comment (eventualbuddha) [reply libs/image-utils/src/image_data.ts:127]: Ah yep. Will remove 👍

File: apps/scan/backend/src/scanners/pdi/state_machine.ts (hunk 1 above)
Comment (eventualbuddha) [reply apps/scan/backend/src/scanners/pdi/state_machine.ts:47]: I do. I think we're going to want to be able to get this timing information in the future.

**Pull request merged**